### PR TITLE
[enh] Conda recipe update and CI OS version pinning

### PR DESCRIPTION
**Pull request by Alexsandruss**
### Description

Scikit-learn-intelex conda recipe update.

Changes:
- Update content of recipe, building and testing scripts to sync with conda-forge feedstock
- Add recipe building on Linux and Windows to CI
- Pin CI images versions
- Add hint for CMake to get python executable from $PYTHON if set

Conda-forge checks: https://github.com/conda-forge/scikit-learn-intelex-feedstock/pull/38

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.  
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

File: scripts/build_backend.py
Comment (icfaust) [inline scripts/build_backend.py:67]: how come icx-cl? so far my builds with icx seem to work...

File: conda-recipe/run_test.sh
Comment (icfaust) [inline conda-recipe/run_test.sh:29]:
```suggestion
    echo "run_test.sh did not find the required testing directories"
```

File: conda-recipe/meta.yaml
Comment (icfaust) [inline conda-recipe/meta.yaml:83]: How is this list verified?

File: conda-recipe/meta.yaml
Comment (icfaust) [inline conda-recipe/meta.yaml:39]: When I search sklearnex for ```DPCPPROOT``` I now only find it in the conda-recipe (no longer in build steps). Is this something we need, and if so, can we give a better description somewhere to a user?

File: conda-recipe/conda_build_config.yaml
Comment (icfaust) [inline conda-recipe/conda_build_config.yaml:20]: Private CI Visual Studio doesn't utilize this compiler, though public CI does. I think it worth having a discussion how we verify this support on private CI-side thoughts? @homksei @ethanglaser

File: .ci/pipeline/nightly.yml
Comment (icfaust) [inline .ci/pipeline/nightly.yml:37]: This should be in a separate PR.  Forcing the VM OS for Coverity/Jupyter checks doesn't seem related and isn't mentioned in the description. Doesn't seem related to the conda build. (Will fast track review when split off).

File: .ci/pipeline/build-and-test-win.yml
Comment (icfaust) [inline .ci/pipeline/build-and-test-win.yml:54]: Interesting. For my edification, could you explain why this change is necessary (may impact a PR of mine)

**Review by icfaust — COMMENTED** [review]
Overall pretty useful addition. I would beef up the description, separate out some of the other changes, and possibly drop sklearn 1.1 or 1.2 from the standard test matrix.  The reason being is that by adding these two recipes, we are starting to really push our luck on azure pipeline runners. I'll let others weigh in on this @samir-nasibli @ethanglaser .

File: .ci/pipeline/build-and-test-win.yml
Comment (Alexsandruss) [reply .ci/pipeline/build-and-test-win.yml:54]: It is to save path compatibility for custom path and non-set path. Our CI runs tests from `s` directory while conda recipe builder - from default path.

File: .ci/pipeline/nightly.yml
Comment (Alexsandruss) [reply .ci/pipeline/nightly.yml:37]: It is actually mentioned: `- Pin CI images versions`.
As CI has update in this PR anyway, I thought it's convenient to include pinning as no actual change in used images is happening.

File: scripts/build_backend.py
Comment (Alexsandruss) [reply scripts/build_backend.py:67]: No actual reason or difference, it was added during debugging. I'm going to remove it.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:39]: It is needed to pass external DPC compiler for `conda-build`, I added docs section for this.

File: conda-recipe/conda_build_config.yaml
Comment (Alexsandruss) [reply conda-recipe/conda_build_config.yaml:20]: I don't see any need to re-specification of `vs` version until we are enable building of conda package in internal CI: azure CI images and most of users use vs2022 anyway

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:83]: What do you mean as the verification in this case?

**Review by bdmoore1 — APPROVED** [review]

File: conda-recipe/meta.yaml
Comment (icfaust) [reply conda-recipe/meta.yaml:83]: I guess this looks to be another list similar to dependencies-dev, just worried about the maintenance of it. Not a big deal, just something I was thinking about, and not sure how it was getting maintained vs dependencies-dev.

File: conda-recipe/conda_build_config.yaml
Comment (icfaust) [reply conda-recipe/conda_build_config.yaml:20]: Seems reasonable.  I will still poke at private CI people as to why we are using an outdated vs version since most are using 2022 as you mentioned.

File: .ci/pipeline/nightly.yml
Comment (icfaust) [reply .ci/pipeline/nightly.yml:37]: I went on ahead and changed the title to make sure this was also there, I think splitting off to a separate PR is a waste of time, just as long as it doesn't get lost in the shuffle.

File: conda-recipe/run_test.bat
Comment (icfaust) [inline conda-recipe/run_test.bat:46]: Does private CI use run_test.bat? I know that the public windows doesn't build the dpc portion and test against dpctl/dpnp so some of these tests aren't verified against public CI.  May be worth running an /intelci: run and checking the testing. It looks like it was deselected in your #1115 but there isn't a comment as to why.

**Review by icfaust — COMMENTED** [review]
Looks to be some sort of windows python conda issue. Your previous responses were sufficient for me. I found one issue with a re-enablement that may require running private CI, maybe some clarity could explain it away as an alternate solution. If we solve these small issues, I think it would be ready.

**Comment by samir-nasibli** [comment]
@Alexsandruss please rebase your branch.

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [reply conda-recipe/run_test.bat:46]: Private CI doesn't use run_test.bat, and I don't remember reason why deselection was introduced. Anyway, it works now.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:83]: I'm afraid there will be at least two files with dependencies: one for conda-recipe and others for non-conda build. Currently, all of them are tested in CIs.

File: INSTALL.md
Comment (icfaust) [inline INSTALL.md:221]:
```suggestion
* any `conda` distribution (`miniforge` is recommended)
```
spelling error in ```preferable```

File: conda-recipe/meta.yaml
Comment (icfaust) [inline conda-recipe/meta.yaml:69]: Only dpc support on Linux (if I am understanding this correct) for the conda meta?

**Review by icfaust — COMMENTED** [review]
So far its okay to have Ubuntu-22, but we should have a ticket in the backlog to check for the Ubuntu 24 drop for azure pipelines.  Or do we have renovatebot checking for this?

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:69]: Yes, only linux, there are unresolved issues with dpc building on Windows using conda-forge dpcpp compiler package.

File: INSTALL.md
Comment (Alexsandruss) [reply INSTALL.md]: Corrected.

File: conda-recipe/meta.yaml
Comment (icfaust) [inline conda-recipe/meta.yaml:75]: Actually now that I think of it, is ```tests:``` evaluated at all in ci?  I see that the azure pipelines job will build it, but do we try to evaluate the testing side of meta.yml?  If no, then I wouldn't do the full testing suite. To go off my previous review comment, I think model builders would best test the ```requres:``` list.

File: INSTALL.md
Comment (icfaust) [inline INSTALL.md:229]: Please add commentary here about the divergence of windows and linux for dpc support. Also maybe MPI depending on the discussion in the other review comment.

File: conda-recipe/meta.yaml
Comment (icfaust) [reply conda-recipe/meta.yaml:84]: Seems reasonable, I guess that is a bit the reason to add a specific CI job for it. thank you,

File: conda-recipe/meta.yaml
Comment (icfaust) [inline conda-recipe/meta.yaml:62]: I don't know much about the mpi-side of things. You have proposed switching the linux support from ```impi``` to ```mpich```.  Could you check with @ethanglaser about these things (out of my knowledge).  Also mpi support for windows is dropped (and is in a comment that may need to be deleted). Is this something to be readded in the future? Also, not to be annoying: Given that dpcpp_cpp_rt was on the ```run:``` side of things, did we have a regression in windows support via this at some point (i.e. did we support dpc building via conda for windows before, or was it just not properly tested, and was erroneously added to ```run:```)?

**Review by icfaust — COMMENTED** [review]
Bit of a deeper-dive review. I think with these comments addressed, you'll get my approval.

File: .ci/pipeline/ci.yml
Comment (samir-nasibli) [inline .ci/pipeline/ci.yml:80]: Please address message from the logs.
The build conda recipe pipeline shows warning 
```
Free memory is lower than 5%; Currently used: 95.03%
```

File: .ci/pipeline/ci.yml
Comment (samir-nasibli) [inline .ci/pipeline/ci.yml:85]: It make sense build with no tests (or just run package import test) and run functional tests separately in other step. What do you think?

**Review by samir-nasibli — COMMENTED** [review]
initial review done.

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:85]: Doesn't make sense: more steps and script lines in CI, reading logs doesn't become easier anyway.

File: conda-recipe/meta.yaml
Comment (napetrov) [reply conda-recipe/meta.yaml:62]: there should be split between GPU mpi support and CPU mpi support.  As well as Intel build of scikit vs public one. 

Overall mpich and mpi are compatibe

Comment (ethanglaser) [reply conda-recipe/meta.yaml:62]: It's a good point - is there a reason we have mpich here instead of impi_rt/impi-devel?

File: conda-recipe/run_test.sh
Comment (ethanglaser) [inline conda-recipe/run_test.sh:1]: This looks much better, thanks for doing this

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml]: I updated conda recipe to sync with requirements-test.txt.
dpctl/dpnp are not going to be included since they are absent from conda-forge channel.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:86]: I removed `daal4py/sklearn`, it's not necessary at this point. Model builders tests are contained in `tests` and included.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml:73]: `conda-build` seems to prefer `run_test.bat / run_test.sh` over test commands in recipe, so full testing is included in conda-recipe CI by default.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml]: I reverted building and runtime deps to use impi.
Feedstock builds of sklearnex never included dpc part, so there is no regression in support on Windows.

File: INSTALL.md
Comment (Alexsandruss) [reply INSTALL.md:229]: I added OS specific details for DPC compiler in INSTALL.md.

File: conda-recipe/meta.yaml
Comment (icfaust) [reply conda-recipe/meta.yaml:73]: @Alexsandruss thank you for the note, this is what impacts #2107 and #2104 @david-cortes-intel

File: .ci/pipeline/ci.yml
Comment (icfaust) [inline .ci/pipeline/ci.yml:101]: So hardcoding the numpy versions is going to increase maintenance burden, is there a reason why they are pinned, and why pinned to < 2.0?

File: .ci/pipeline/ci.yml
Comment (icfaust) [inline .ci/pipeline/ci.yml:91]: Where is the run_test.sh occur here? (sorry if its a dumb question, I don't know much about the conda building testing standard procedures) (given https://github.com/intel/scikit-learn-intelex/pull/1992/files#r1801556134)

**Review by icfaust — COMMENTED** [review]
Just some last questions from me. Thanks for the updates based off the previous comments.

**Comment by david-cortes-intel** [comment]
Question: are the commands in `meta.yaml` for executing tests actually needed if it will end up calling the `run_test.sh/bat` scripts in the end?

File: .ci/pipeline/ci.yml
Comment (samir-nasibli) [reply .ci/pipeline/ci.yml:91]: @icfaust https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#test-section
If this section exists or if there is a `run_test.[py,pl,sh,bat,r]` file in the recipe, the package is installed into a test environment after the build is finished and the tests are run there.

About dpnp/dpctl test skipping. I think it should be noted somewhere and tracked when it could be enabled.

File: .ci/pipeline/ci.yml
Comment (samir-nasibli) [inline .ci/pipeline/ci.yml:70]: Any justification for pinning such numpy versions?

File: .ci/pipeline/ci.yml
Comment (samir-nasibli) [inline .ci/pipeline/ci.yml:90]: Since there an option only build without testing I suggest add `test` in the name of the job for better description of the flow.
```suggestion
    displayName: "Conda build and test"
```

File: .ci/scripts/gen_release_jobs.py
Comment (samir-nasibli) [inline .ci/scripts/gen_release_jobs.py:27]: I think it worth to add some comment for better understanding and justification why not the latest versions is used.

File: conda-recipe/meta.yaml
Comment (samir-nasibli) [inline conda-recipe/meta.yaml:20]: I suggest to use git env variables here
https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#git-environment-variables

File: conda-recipe/meta.yaml
Comment (samir-nasibli) [inline conda-recipe/meta.yaml:63]: Maybe for the current iteration it is fine, but for a such project like `scikit-learn-intelex` I expect that it will be enhanced by using `pyproject.toml`, where requirements defined in one place. It will require to update `pyproject.toml` file, `[build-system]` table is recommended https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#writing-your-pyproject-toml

File: conda-recipe/conda_build_config.yaml
Comment (david-cortes-intel) [reply conda-recipe/conda_build_config.yaml:18]: And actually on second thought: since these are **build-time** dependencies rather than runtime, perhaps would be better to switch them to numpy>=2, since a module built with numpy>=2 is usable with a runtime numpy<2, but not the other way around.

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:101]: Current numpy handling is similar to `dependencies-dev` and `requirements-test.txt`: lowest possible numpy version is set for building and `{{ pin_compatible(numpy)) }}` installs highest compatible version for testing. The only difference is `conda-build` ignores `numpy>=2.0`

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:91]: I added note about dpnp/dpctl.

File: conda-recipe/conda_build_config.yaml
Comment (Alexsandruss) [reply conda-recipe/conda_build_config.yaml:18]: I would like to leave numpy installation behavior similar to usual testing: https://github.com/intel/scikit-learn-intelex/pull/1992#discussion_r1804754817

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:70]: https://github.com/intel/scikit-learn-intelex/pull/1992#discussion_r1804754817

File: conda-recipe/conda_build_config.yaml
Comment (david-cortes-intel) [reply conda-recipe/conda_build_config.yaml:18]: @Alexsandruss Wouldn't that leave parts of the conda package broken if a user first install scikit-learn-intelex and then later on upgrades numpy to a version >= 2?

File: .ci/scripts/gen_release_jobs.py
Comment (Alexsandruss) [reply .ci/scripts/gen_release_jobs.py:29]: Added.

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:36]: Is this one used anywhere when running on linux? I understood it was meant for the MPI invocation on windows only.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:18]: Shouldn't this say "%cd%" since it's windows?

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:22]:
```suggestion
IF NOT DEFINED PYTHON (set "PYTHON=python")
```

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:31]: Wouldn't `%1%` be interpreted as an environment variable instead of CLI argument?

File: conda-recipe/run_test.sh
Comment (Alexsandruss) [reply conda-recipe/run_test.sh:36]: For `${PYTHON} -c "from sklearnex import patch_sklearn; patch_sklearn()"` at least. It might be required for some additional tests later, so no need to remove it.

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [reply conda-recipe/run_test.bat:31]: `%1%` and `%1` works the same way here as I saw.

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [reply conda-recipe/run_test.bat:22]: Done.

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [reply conda-recipe/run_test.bat:18]: I added mention of both.

File: INSTALL.md
Comment (napetrov) [inline INSTALL.md:226]: can we point to general VS/compiler requirements so we would have single place to track/update?

Comment (Alexsandruss) [reply INSTALL.md:226]: Not really in case of conda recipe, vs2022 is default for this repo and included in CI Windows image, vs2019 - for conda-forge feedstock, and build config change is required to set another version and there is no guarantee it will work as expected (but it is unlikely someone will use version other than 2019 or 2022).

I will add mention how to change compiler for conda-build on Win.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [reply conda-recipe/run_test.bat:31]: Isn't that just a coincidence due to the way you placed the variable?

For example, you can try a small script like this:
```batch
echo %1% %2%
```

and another like this:
```batch
echo %1 %2
```

then call each with arguments like these:
```batch
file.bat arg1 arg2
```

and you'd see that once there is more than one argument or they are somewhere where an extra `%` wouldn't be ignored, it's not the same to use `%1%` instead of `%1`.

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [reply conda-recipe/run_test.sh:36]: Is `${PYTHON}` ever set to something other than "python" when executing the .sh script? If not, would be nice to add it as a comment in the file.

File: conda-recipe/meta.yaml
Comment (Alexsandruss) [reply conda-recipe/meta.yaml]: Added

File: conda-recipe/run_test.sh
Comment (Alexsandruss) [reply conda-recipe/run_test.sh:36]: It's used by `conda-build`: https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#environment-variables-set-during-the-build-process

File: conda-recipe/run_test.bat
Comment (Alexsandruss) [reply conda-recipe/run_test.bat:31]: Yes, it works as you described, I'll fix it

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:44]: Why would this one have an `echo` entry while the `echo` entries of others were removed?

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:50]: I think it'd be quite helpful when inspecting failures to have this one execute before the MPI one, like it was before.

That way, if something fails and execution stops during the MPI part, one would then know whether the problem was with MPI specifically or with the tests in general.

**Review by david-cortes-intel — APPROVED** [review]

**Comment by icfaust** [comment]
@Alexsandruss This was merged with an error in the WindowsCondaEnv Python3.12_Sklearn1.5 step

**Comment by david-cortes-intel** [comment]
> @Alexsandruss This was merged with an error in the WindowsCondaEnv Python3.12_Sklearn1.5 step

The error appears to be sporadic though.

**Pull request merged**
